src/main/java/com/appsflyer/otelawsmetrics/OtelMetricPublisher.java

```diff
@@ -0,0 +1,289 @@
+        }
+    }
+
+    private void recordMetrics(Map<String, MetricRecord<?>> metricsMap,
+                               Attributes attributes,
+                               Map<String, MetricStrategy> metricStrategies) {
+        for (Map.Entry<String, MetricStrategy> entry : metricStrategies.entrySet()) {
+            MetricRecord<?> metricRecord = metricsMap.get(entry.getKey());
+            if (metricRecord != null) {
+                entry.getValue().record(metricRecord, attributes);
+            }
+        }
+    }
+
+    private void processPerRequestMetrics(MetricCollection requestMetrics) {
+        Map<String, MetricRecord<?>> metricsMap = extractMetrics(requestMetrics);
+
+        // Extract attributes for per-request metrics
+        String operationName = getStringMetricValue(metricsMap, CoreMetric.OPERATION_NAME.name());
+        boolean isSuccess = getBooleanMetricValue(metricsMap, CoreMetric.API_CALL_SUCCESSFUL.name());
+        int retryCount = getIntMetricValue(metricsMap, CoreMetric.RETRY_COUNT.name());
+        Attributes attributes = toPerRequestAttributes(operationName, isSuccess, retryCount);
+
+        // Report per-request metrics
+        recordMetrics(metricsMap, attributes, perRequestMetrics);
+
+        // Process per-attempt metrics
+        for (MetricCollection attemptMetrics : requestMetrics.children()) {
+            processPerAttemptMetrics(attemptMetrics, attributes);
+        }
+    }
+
+    private void processPerAttemptMetrics(MetricCollection attemptMetrics, Attributes parentAttributes) {
+        Map<String, MetricRecord<?>> metricsMap = extractMetrics(attemptMetrics);
+
+        // Extract ErrorType if present
+        String errorType = getStringMetricValue(metricsMap, CoreMetric.ERROR_TYPE.name());
+
+        // Build attributes including attempt number and error type
+        Attributes attributes = toAttemptAttributes(parentAttributes, errorType);
+
+        // Report per-attempt metrics
+        recordMetrics(metricsMap, attributes, perAttemptMetrics);
+
+        // Process HTTP metrics
+        for (MetricCollection httpMetricsCollection : attemptMetrics.children()) {
+            processHttpMetrics(httpMetricsCollection, attributes);
+        }
+    }
+
+    private void processHttpMetrics(MetricCollection httpMetricsCollection, Attributes parentAttributes) {
+        Map<String, MetricRecord<?>> metricsMap = extractMetrics(httpMetricsCollection);
+
+        // Extract HTTP status code
+        int httpStatusCode = getIntMetricValue(metricsMap, HttpMetric.HTTP_STATUS_CODE.name());
+        Attributes attributes = toHttpAttributes(parentAttributes, httpStatusCode);
+
+        // Report HTTP metrics
+        recordMetrics(metricsMap, attributes, httpMetrics);
+    }
+
+    private Map<String, MetricRecord<?>> extractMetrics(MetricCollection metricCollection) {
+        Map<String, MetricRecord<?>> metricMap = new HashMap<>();
```

Comment:
You can express this whole function using Stream API very elegantly -

```java
metricCollection.stream.collect(Collectors.toMap(metricRecord -> metricRecord.metric().name(), Function.identity()))
```

Reply:
The `extractMetrics` method is called 3 times for each `MetricPublisher.publish` call.  
For a high-throughput application, the overhead of calling `.stream()` on a collection with fewer than 100 elements is non-negligible.  
See [this StackOverflow reply](https://stackoverflow.com/a/41823317) with [source code](https://pastebin.com/TzWj1MH1) you can verify on your machine.